server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java

```diff
@@ -1621,23 +1621,34 @@ public void validateIndexSettings(String indexName, final Settings settings, fin
         }
     }
 
-    List<String> getIndexSettingsValidationErrors(final Settings settings, final boolean forbidPrivateIndexSettings) {
+    List<String> getIndexSettingsValidationErrors(
+        final Settings settings,
+        @Nullable Settings systemProvided,
+        final boolean forbidPrivateIndexSettings
+    ) {
         List<String> validationErrors = validateIndexCustomPath(settings, env.sharedDataDir());
         if (forbidPrivateIndexSettings) {
-            validationErrors.addAll(validatePrivateSettingsNotExplicitlySet(settings, indexScopedSettings));
+            validationErrors.addAll(validatePrivateSettingsNotExplicitlySet(settings, systemProvided, indexScopedSettings));
         }
         return validationErrors;
     }
 
-    private static List<String> validatePrivateSettingsNotExplicitlySet(Settings settings, IndexScopedSettings indexScopedSettings) {
+    private static List<String> validatePrivateSettingsNotExplicitlySet(
+        Settings settings,
+        @Nullable Settings systemProvided,
+        IndexScopedSettings indexScopedSettings
+    ) {
         List<String> validationErrors = new ArrayList<>();
         for (final String key : settings.keySet()) {
             final Setting<?> setting = indexScopedSettings.get(key);
             if (setting == null) {
                 assert indexScopedSettings.isPrivateSetting(key) : "expected [" + key + "] to be private but it was not";
-            } else if (setting.isPrivateIndex()) {
-                validationErrors.add("private index setting [" + key + "] can not be set explicitly");
-            }
+            } else if (setting.isPrivateIndex()
+                // System-provided settings are always allowed to configure private settings.
+                // These are typically coming from an IndexSettingProvider.
+                && (systemProvided == null || settings.get(key).equals(systemProvided.get(key)) == false)) {
+                    validationErrors.add("private index setting [" + key + "] can not be set explicitly");
+                }
```

Comment:
Would it be better to leave the `if` statements alone and instead do something like:

```java
for (final String key : settings.filter(s -> systemProvided.hasValue(s.getKey()) == false)).keySet()) {
    …
}
```

to filter the system-provided settings out of the validation list at the beginning?